File: mobile/native/ios/InfraAdvisorMobile/APIClient.swift
import Foundation

/// Concrete delegate required by Datadog's URLSession swizzling. The delegate does not need
/// application callbacks; Datadog injects the lifecycle methods it observes at runtime.
final class InfraAdvisorURLSessionDelegate: NSObject, URLSessionDataDelegate {}

protocol APIClientProtocol {
    func login(email: String, password: String) async throws -> LoginResponse
    func models(backend: Backend) async throws -> ModelsResponse
    func listConversations(token: String) async throws -> [ConversationSummary]
    func conversation(token: String, id: String) async throws -> ConversationDetail
    func createConversation(token: String, title: String, model: String, backend: Backend) async throws -> ConversationSummary
    func query(token: String, prompt: String, sessionID: String, model: String, backend: Backend, userID: String, conversationID: String) async throws -> QueryResponse
}

enum APIClientError: LocalizedError, Equatable {
    case invalidResponse
    case http(status: Int, message: String)
    case decoding

    var errorDescription: String? {
        switch self {
        case .invalidResponse: return "The server returned an invalid response."
        case let .http(_, message): return message
        case .decoding: return "The server response could not be read."
        }
    }
}

final class APIClient: APIClientProtocol {
    private let baseURL: URL
    private let session: URLSession
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    init(baseURL: URL, session: URLSession) {
        self.baseURL = baseURL
        self.session = session
    }

    func login(email: String, password: String) async throws -> LoginResponse {
        try await post(path: "/auth/login", body: ["email": email, "password": password], token: nil)
    }

    func models(backend: Backend) async throws -> ModelsResponse {
        try await get(path: backend.apiPrefix + "/models", token: nil)
    }

    func listConversations(token: String) async throws -> [ConversationSummary] {
        let response: ConversationListResponse = try await get(path: "/api/conversations", token: token)
        return response.conversations
    }

    func conversation(token: String, id: String) async throws -> ConversationDetail {
        try await get(path: "/api/conversations/\(id)", token: token)
    }

    func createConversation(token: String, title: String, model: String, backend: Backend) async throws -> ConversationSummary {
        try await post(path: "/api/conversations", body: ConversationCreateRequest(title: title, model: model, backend: backend), token: token)
    }

    func query(token: String, prompt: String, sessionID: String, model: String, backend: Backend, userID: String, conversationID: String) async throws -> QueryResponse {
        var headers = ["Content-Type": "application/json", "X-Session-ID": sessionID, "X-User-ID": userID, "X-Conversation-ID": conversationID]
        headers["Authorization"] = "Bearer \(token)"
        return try await request(path: backend.apiPrefix + "/query", method: "POST", body: encoder.encode(QueryRequest(query: prompt, sessionID: sessionID, model: model)), headers: headers, timeout: 90)
    }

    private func get<Response: Decodable>(path: String, token: String?) async throws -> Response {
        var headers: [String: String] = [:]
        if let token { headers["Authorization"] = "Bearer \(token)" }
        return try await request(path: path, method: "GET", body: nil, headers: headers)
    }

    private func post<Response: Decodable, Body: Encodable>(path: String, body: Body, token: String?) async throws -> Response {
        var headers = ["Content-Type": "application/json"]
        if let token { headers["Authorization"] = "Bearer \(token)" }
        return try await request(path: path, method: "POST", body: encoder.encode(body), headers: headers)
    }

    private func request<Response: Decodable>(path: String, method: String, body: Data?, headers: [String: String], timeout: TimeInterval = 60) async throws -> Response {
        // This request must run through the delegate-backed session created after
        // URLSessionInstrumentation.enable. URLSession.shared is not used because Datadog
        // requires a concrete delegate class to observe resources and inject trace context.
        var request = URLRequest(url: baseURL.appending(path: path))
        request.httpMethod = method
        request.timeoutInterval = timeout
        headers.forEach { request.setValue($0.value, forHTTPHeaderField: $0.key) }
        request.httpBody = body

        // Privacy boundary: request/response bodies and Authorization are used only by the
        // HTTP client. They are never copied into RUM attributes or span tags.
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw APIClientError.invalidResponse }
        guard (200..<300).contains(http.statusCode) else {
            let body = try? decoder.decode(APIErrorBody.self, from: data)
            let fallback = HTTPURLResponse.localizedString(forStatusCode: http.statusCode)
            throw APIClientError.http(status: http.statusCode, message: body?.detail ?? body?.message ?? "Request failed: \(fallback) (\(http.statusCode))")
        }
        do { return try decoder.decode(Response.self, from: data) }
        catch { throw APIClientError.decoding }
    }
}

File: mobile/native/android/app/src/main/java/dev/kyletaylor/infraadvisor/mobile/ApiClient.java
package dev.kyletaylor.infraadvisor.mobile;

import android.content.Context;
import com.android.volley.RequestQueue;
import com.android.volley.Request;
import com.android.volley.Response;
import com.android.volley.toolbox.Volley;
import dev.kyletaylor.infraadvisor.mobile.api.ApiParsers;
import dev.kyletaylor.infraadvisor.mobile.model.LoginResponse;
import dev.kyletaylor.infraadvisor.mobile.model.QueryResponse;
import dev.kyletaylor.infraadvisor.mobile.model.ConversationDetail;
import dev.kyletaylor.infraadvisor.mobile.model.ConversationSummary;
import dev.kyletaylor.infraadvisor.mobile.observability.ObservedJsonRequest;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;
import org.json.JSONException;
import org.json.JSONObject;

public final class ApiClient {
    public interface Result<T> { void success(T value); void failure(Exception error); }
    public static final String BACKEND_PYTHON = "python";
    public static final String BACKEND_DOTNET = "dotnet";
    private static final int STANDARD_TIMEOUT_MS = 15_000;
    private static final int QUERY_TIMEOUT_MS = 90_000;

    public static final class ModelOptions {
        public final List<String> models;
        public final String defaultModel;
        ModelOptions(List<String> models, String defaultModel) {
            this.models = models;
            this.defaultModel = defaultModel;
        }
    }
    private final String baseUrl;
    private final RequestQueue queue;

    public ApiClient(Context context, String baseUrl) {
        this.baseUrl = baseUrl.replaceAll("/+$", "");
        this.queue = Volley.newRequestQueue(context.getApplicationContext());
    }

    public void login(String email, String password, Result<LoginResponse> result) {
        try {
            JSONObject body = new JSONObject().put("email", email).put("password", password);
            enqueue(Request.Method.POST, "/auth/login", body, Collections.emptyMap(), STANDARD_TIMEOUT_MS,
                    json -> parseLogin(json, result), result::failure);
        } catch (JSONException error) { result.failure(error); }
    }

    public void query(String token, String prompt, String sessionId, String model, String backend, String userId,
                      String conversationId,
                      Result<QueryResponse> result) {
        try {
            JSONObject body = new JSONObject().put("query", prompt).put("session_id", sessionId).put("model", model);
            Map<String, String> headers = new HashMap<>();
            headers.put("Authorization", "Bearer " + token);
            headers.put("X-Session-ID", sessionId);
            headers.put("X-User-ID", userId);
            headers.put("X-Conversation-ID", conversationId);
            enqueue(Request.Method.POST, apiPath(backend, "/query"), body, headers, QUERY_TIMEOUT_MS,
                    json -> parseQuery(json, result), result::failure);
        } catch (JSONException error) { result.failure(error); }
    }

    public void models(String backend, Result<ModelOptions> result) {
        enqueue(Request.Method.GET, apiPath(backend, "/models"), null, Collections.emptyMap(), STANDARD_TIMEOUT_MS,
                json -> {
                    List<String> models = new ArrayList<>();
                    org.json.JSONArray values = json.optJSONArray("models");
                    if (values != null) {
                        for (int index = 0; index < values.length(); index++) {
                            String value = values.optString(index, "");
                            if (!value.isEmpty()) models.add(value);
                        }
                    }
                    if (models.isEmpty()) models.add("gpt-4.1-mini");
                    result.success(new ModelOptions(models, json.optString("default", models.get(0))));
                }, result::failure);
    }

    public void listConversations(String token, Result<List<ConversationSummary>> result) {
        enqueue(Request.Method.GET, "/api/conversations", null, authHeaders(token), STANDARD_TIMEOUT_MS,
                json -> {
                    try { result.success(ApiParsers.conversations(json)); }
                    catch (JSONException error) { result.failure(error); }
                }, result::failure);
    }

    public void conversation(String token, String id, Result<ConversationDetail> result) {
        enqueue(Request.Method.GET, "/api/conversations/" + id, null, authHeaders(token), STANDARD_TIMEOUT_MS,
                json -> {
                    try { result.success(ApiParsers.conversation(json)); }
                    catch (JSONException error) { result.failure(error); }
                }, result::failure);
    }

    public void createConversation(String token, String title, String model, String backend,
                                   Result<ConversationSummary> result) {
        try {
            JSONObject body = new JSONObject().put("title", title).put("model", model).put("backend", backend);
            enqueue(Request.Method.POST, "/api/conversations", body, authHeaders(token), STANDARD_TIMEOUT_MS,
                    json -> {
                        try { result.success(ApiParsers.conversationSummary(json)); }
                        catch (JSONException error) { result.failure(error); }
                    }, result::failure);
        } catch (JSONException error) { result.failure(error); }
    }

    private void enqueue(int method, String path, JSONObject body, Map<String, String> headers, int timeoutMs,
                         Response.Listener<JSONObject> listener, Response.ErrorListener errors) {
        queue.add(new ObservedJsonRequest(method, baseUrl + path, body, headers, timeoutMs, listener, errors));
    }

    static String apiPath(String backend, String route) {
        return (BACKEND_DOTNET.equals(backend) ? "/api-dotnet" : "/api") + route;
    }

    private static Map<String, String> authHeaders(String token) {
        Map<String, String> headers = new HashMap<>();
        headers.put("Authorization", "Bearer " + token);
        return headers;
    }

    private static void parseLogin(JSONObject json, Result<LoginResponse> result) {
        try { result.success(ApiParsers.login(json)); } catch (JSONException error) { result.failure(error); }
    }
    private static void parseQuery(JSONObject json, Result<QueryResponse> result) {
        try { result.success(ApiParsers.query(json)); } catch (JSONException error) { result.failure(error); }
    }
}
